### PR TITLE
AdView: support multi imps request & formattype tag for bid response

### PR DESCRIPTION
**Pull request by AdviewOpen**
1. support multi imps from request.
2. add formattype tag for bid response, use this to handle resp bid-type.

### CODE REVIEW

File: adapters/adview/adview.go [hunk 1]
```diff
@@ -157,3 +178,44 @@ func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType,
...
+	for _, imp := range reqImps {
+		if bid.ImpID == imp.ID {
+			if bid.Ext != nil {
+				var bidExt *adviewBidExt
+				err := json.Unmarshal(bid.Ext, &bidExt)
+				if err == nil {
+					return getBidType(bidExt), nil
+				} else {
+					return mediaType, err
+				}
```
Comment (onkarvhanumante) [inline adapters/adview/adview.go:193]: code path is not covered. should add json test for better coverage

File: adapters/adview/adview.go [hunk 2]
```diff
@@ -157,3 +178,44 @@ func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType,
...
+	// setting "banner" as the default bid type
+	bidType := openrtb_ext.BidTypeBanner
+	if bidExt != nil {
+		switch bidExt.BidType {
+		case 0:
+			bidType = openrtb_ext.BidTypeBanner
+			//fmt.Println("==== [adview] getBidType(): ==> banner ===")
+		case 1:
+			bidType = openrtb_ext.BidTypeVideo
+			//fmt.Println("==== [adview] getBidType(): ==> video ===")
```
Comment (onkarvhanumante) [inline adapters/adview/adview.go:212]: seems debug left over. should remove commented fmt lines on line 209 and 212

File: adapters/adview/adview.go [hunk 3]
```diff
@@ -117,12 +136,14 @@ func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.R
...
-	bidResponse.Currency = "USD" //we just support USD for resp
+	//we just support USD for resp
+	bidResponse.Currency = "USD"
 
 	var errors []error
 	for _, seatBid := range response.SeatBid {
 		for i, bid := range seatBid.Bid {
-			bidType, err := getMediaTypeForImp(bid.ImpID, request.Imp)
+			bidType, err := getMediaTypeForBid(bid, request.Imp)
+			//bidType, err := getMediaTypeForImp(bid.ImpID, request.Imp)
```
Comment (onkarvhanumante) [inline adapters/adview/adview.go:146]: seems debug leftover. should remove code comment on line 146

File: adapters/adview/adview.go (hunk 1 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:192]: thx, and I add a test script banner-app-resp-no-formattype.json, and also adview ad server always deliver 'formattype' in ad response, so formattype error cases may not be occurred.

File: adapters/adview/adview.go (hunk 2 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go]: removed.

File: adapters/adview/adview.go (hunk 3 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go]: removed & thx.

**Comment by AdviewOpen** [comment]
Hi, I committed the changes, pls check it again.

File: adapters/adview/adview.go [hunk 4]
```diff
@@ -157,3 +177,42 @@ func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType,
...
+		switch bidExt.BidType {
+		case 0:
+			bidType = openrtb_ext.BidTypeBanner
+		case 1:
+			bidType = openrtb_ext.BidTypeVideo
+		case 2:
+			bidType = openrtb_ext.BidTypeNative
+		default:
+			// default is banner
+			bidType = openrtb_ext.BidTypeBanner
```
Comment (onkarvhanumante) [inline adapters/adview/adview.go:214]: should error if none of the cases match

Comment (AdviewOpen) [reply adapters/adview/adview.go:214]: I have confirmed that adview server side just support banner,video&native, there is no other value will be returned for response, so default case will not occur. thanks.

Comment (onkarvhanumante) [reply adapters/adview/adview.go:214]: Adding default case and returning error will help to identify case if adview server is not returing media type. Therefore will insist to return error if none of the switch case match

Comment (AdviewOpen) [reply adapters/adview/adview.go]: Ok, agree with you and i committed the changes. pls check it again.

**Comment by AdviewOpen** [comment]
new changes for default cases error process in getBidType. pls check it & thx.

**Review by onkarvhanumante — APPROVED** [review]

File: adapters/adview/adview.go [hunk 5]
```diff
@@ -157,3 +177,41 @@ func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType,
...
+	}
+	return mediaType, nil
+}
+
+// getBidType returns the bid type specified in the response bid.ext
+func getBidType(bidExt *adviewBidExt) (openrtb_ext.BidType, error) {
+	// setting "banner" as the default bid type
+	bidType := openrtb_ext.BidTypeBanner
+	if bidExt != nil {
+		switch bidExt.BidType {
```
Comment (Sonali-More-Xandr) [inline adapters/adview/adview.go:205]: Please consider implementing a pattern where the adapter server sets the [MType](https://github.com/prebid/openrtb/blob/main/openrtb2/bid.go#L334) field in the response to accurately determine the media type for the impression.

Comment (AdviewOpen) [reply adapters/adview/adview.go:205]: Sorry, do you mean in line 203, bidType set to banner default type is not suitable, so I should return error if no matched bidtype, right?
Or you mean I should set Mtype such as "seatBid.Bid[i].MType = openrtb2.MarkupBanner" in MakeBids()?
Thank you.

Comment (AdviewOpen) [reply adapters/adview/adview.go:205]: @Sonali-More-Xandr So, you mean is I can use the following codes to get the bid type is enough?
func getMediaTypeForBid(bid openrtb2.Bid) (openrtb_ext.BidType, error) {
	switch bid.MType {
	case openrtb2.MarkupBanner:
		return openrtb_ext.BidTypeBanner, nil
	case openrtb2.MarkupVideo:
		return openrtb_ext.BidTypeVideo, nil
	case openrtb2.MarkupNative:
		return openrtb_ext.BidTypeNative, nil
	default:
		return "", fmt.Errorf("Unable to fetch mediaType in: %s", bid.ImpID)
	}
}

**Comment by AdviewOpen** [comment]
Based on reviews, and I got the confirmation that adview server already support mtype in response, after test, i change getMediaTypeForBid(), pls review it. 
Thanks.

File: adapters/adview/adview.go [hunk 6]
```diff
@@ -157,3 +177,37 @@ func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType,
...
+		return openrtb_ext.BidTypeBanner, nil
+	case openrtb2.MarkupVideo:
+		return openrtb_ext.BidTypeVideo, nil
+	case openrtb2.MarkupAudio:
+		return openrtb_ext.BidTypeAudio, nil
+	case openrtb2.MarkupNative:
+		return openrtb_ext.BidTypeNative, nil
+	default:
+		return "", fmt.Errorf("Unable to fetch mediaType in multi-format: %s", bid.ImpID)
+	}
```
Comment (onkarvhanumante) [inline adapters/adview/adview.go:193]: @AdviewOpen  should also add received `mType` in error case

Comment (AdviewOpen) [reply adapters/adview/adview.go]: Ok, add it.

**Comment by AdviewOpen** [comment]
Hi, I add mtype in error output & removed no used codes. 
pls check it and thx.

File: adapters/adview/adview.go [hunk 7]
```diff
@@ -143,17 +163,17 @@ func (a *adapter) buildEndpointURL(params *openrtb_ext.ExtImpAdView) (string, er
...
-			}
-			return mediaType, nil
-		}
+func getMediaTypeForBid(bid openrtb2.Bid) (openrtb_ext.BidType, error) {
+	switch bid.MType {
+	case openrtb2.MarkupBanner:
+		return openrtb_ext.BidTypeBanner, nil
+	case openrtb2.MarkupVideo:
+		return openrtb_ext.BidTypeVideo, nil
+	case openrtb2.MarkupAudio:
```
Comment (onkarvhanumante) [inline adapters/adview/adview.go:172]: as per `adview` bidder-info yml, audio media type is not supported. should remove audio case

https://github.com/prebid/prebid-server/blob/cd9855403a50d108475dcf66d6d6e734db163be3/static/bidder-info/adview.yaml#L9-L12

Comment (AdviewOpen) [reply adapters/adview/adview.go]: Thanks for the reminder, fixed.

**Comment by AdviewOpen** [comment]
Remove audio type cause not supporting now. thx.

**Review by Sonali-More-Xandr — APPROVED** [review]

**Review by onkarvhanumante — APPROVED** [review]

**Comment by bsardo** [comment]
@AdviewOpen I didn't see a docs PR; can you open one please? Your documentation currently states that your adapter doesn’t support multi impression and can not perform impression splitting, so only the first impression will be delivered.

https://github.com/prebid/prebid.github.io/blob/5b17a6f67927ac955d9bda241bd0ec2c19fb7435/dev-docs/bidders/adview.md?plain=1#L25

**Comment by AdviewOpen** [comment]
We have update the doc PR #5126, Thanks again.

**Pull request merged**